integration-tests/go-sql-server-driver/testdef.go

```diff
@@ -0,0 +1,201 @@
+}
+
+type Server struct {
+	Name       string   `yaml:"name"`
+	Args       []string `yaml:"args"`
+	LogMatches []string `yaml:"log_matches"`
+}
+
+type Query struct {
+	Query      string      `yaml:"query"`
+	Exec       string      `yaml:"exec"`
+	Args       []string    `yaml:"args"`
+	Result     QueryResult `yaml:"result"`
+	ErrorMatch string      `yaml:"error_match"`
+}
+
+type QueryResult struct {
+	Columns []string   `yaml:"columns"`
+	Rows    [][]string `yaml:"rows"`
+}
+
+func ParseTestsFile(path string) (TestDef, error) {
+	contents, err := os.ReadFile(path)
+	if err != nil {
+		return TestDef{}, err
+	}
+	var res TestDef
+	err = yaml.Unmarshal(contents, &res)
+	return res, err
+}
+
+func RunTestsFile(t *testing.T, path string) {
+	def, err := ParseTestsFile(path)
+	require.NoError(t, err)
+	for _, test := range def.Tests {
+		t.Run(test.Name, func(t *testing.T) {
+			u, err := NewDoltUser()
+			require.NoError(t, err)
+			rs, err := u.MakeRepoStore()
+			require.NoError(t, err)
+			repos := make(map[string]Repo)
+			for _, r := range test.Repos {
+				var repo Repo
+				repo, err = rs.MakeRepo(r.Name)
+				require.NoError(t, err)
+				repos[r.Name] = repo
+				for _, f := range r.WithFiles {
+					require.NoError(t, repo.WriteFile(f.Name, f.Contents))
+				}
+			}
+			servers := make(map[string]*SqlServer)
+			for _, sl := range test.Servers {
+				s := sl
+				var server *SqlServer
+				server, err = repos[s.Name].StartSqlServer(s.Args...)
+				require.NoError(t, err)
+				servers[s.Name] = server
+				defer func() {
+					err := server.GracefulStop()
+					require.NoError(t, err)
+					output := string(server.Output.Bytes())
+					for _, a := range s.LogMatches {
+						require.Regexp(t, a, output)
+					}
+				}()
+			}
+			dbs := make(map[string]*sql.DB)
+			for nl, s := range servers {
+				n := nl
+				db, err := s.DB(n)
+				require.NoError(t, err)
+				dbs[n] = db
+				defer func() {
+					dbs[n].Close()
+				}()
+			}
+			for i, c := range test.Conns {
+				db := dbs[c.On]
+				require.NotNilf(t, db, "error in test spec: could not find database %s for connection %d", c.On, i)
+				conn, err := db.Conn(context.Background())
+				require.NoError(t, err)
+				func() {
+					// Do not return this connection to the connection pool.
+					defer conn.Raw(func(any) error {
+						return driver.ErrBadConn
+					})
+					for _, q := range c.Queries {
+						args := make([]any, len(q.Args))
+						for i := range q.Args {
+							args[i] = q.Args[i]
+						}
+						if q.Query != "" {
+							rows, err := conn.QueryContext(context.Background(), q.Query, args...)
+							if q.ErrorMatch != "" {
+								require.Error(t, err)
+								require.Regexp(t, q.ErrorMatch, err.Error())
+							}
+							require.NoError(t, err)
+							defer rows.Close()
+							cols, err := rows.Columns()
+							require.NoError(t, err)
+							require.Equal(t, q.Result.Columns, cols)
+							for _, r := range q.Result.Rows {
+								require.True(t, rows.Next())
```

Comment:
Rather than this assertion inside the loop, it would be better to consume the entire row buffer, then do a simple require.Equals(expected, found) at the end, which will include a length check and have a better error message